### PR TITLE
Add support for command

### PR DESCRIPTION
**Pull request by Fymyte**
Closes: #42
#48
Signed-off-by: Fymyte <pierguill@gmail.com>

### CODE REVIEW

**Comment by NoahTheDuke** [comment]
Any updates on this?

**Comment by Fymyte** [comment]
It should be ready to merge. @vigoux don't have much time

**Comment by vigoux** [comment]
I'll check that when I have time, rebase it and merge it when possible
I'll keep you updated!

**Comment by Fymyte** [comment]
@vigoux I have done the rebase on master

**Comment by vigoux** [comment]
Great, I think it would be good the change the queries and test them according to all this.
Do you think you have time to do it, or do you want me to handle that ?

**Comment by vigoux** [comment]
Great.
BTW you are going to hate me, but I changed the way we add keywords in latest master today:
You now only need to change the `grammar.js` file, and mimic the lines that are already there.

Normally this should avoid trigerring the fuzzing tests for this PR.

**Comment by Fymyte** [comment]
@vigoux I am not sure how to handle the override of certain keywords: `-complete` argument accepts `syntax` for example, which is highlighted as a keyword. I need to override this behavior to match it as a `constant`.
For example:
```vim
command -complete=syntax One echo "one"
```
with this query file `syntax` is highlighted as a keyword

File: grammar.js [hunk 1]
```diff
@@ -395,7 +398,105 @@ module.exports = grammar({
...
+        seq(field('name', '-nargs'), token.immediate('='), field('value', $._command_attribute_nargs_value)),
+        seq(field('name', '-complete'), token.immediate('='), field('value', alias($._command_attribute_completion_behavior, $.behavior))),
+        seq(field('name', '-range'), optional(seq(token.immediate('='), field('value', $._command_attribute_range_value)))),
+        seq(field('name', '-count'), token.immediate('='), field('value', alias(token.immediate(/[0-9]+/), $.integer_literal))),
+        seq(field('name', '-addr'), token.immediate('='), field('value', alias($._command_attribute_address_behavior, $.behavior))),
+        field('name', '-bang'),
+        field('name', '-bar'),
+        field('name', '-register'),
+        field('name', '-buffer'),
+        field('name', '-keepscript'),
```
Comment (vigoux) [inline grammar.js:482]: Could be worth using one of the functions we use for `syntax` here ?

we could use `key_val_arg` ?

File: queries/highlights.scm [hunk 2]
```diff
@@ -134,6 +135,13 @@
   "clear"
 ] @keyword)
 
+;; Command command
+
+(command_attribute name: _ @property)
+(command_attribute
+  value: (behavior
+    (identifier)? @function) @constant)
```
Comment (vigoux) [inline queries/highlights.scm:143]: It is not a always `function` here, right ? Could be an integer or something else.

File: grammar.js [hunk 3]
```diff
@@ -395,7 +398,105 @@ module.exports = grammar({
...
+        alias(token.immediate(/[0-9]+/), $.integer_literal),
+        alias(token.immediate(/%/), $.pattern_multi),
+      ),
+
+    command_attribute: ($) =>
+      choice(
+        seq(field('name', '-nargs'), token.immediate('='), field('value', $._command_attribute_nargs_value)),
+        seq(field('name', '-complete'), token.immediate('='), field('value', alias($._command_attribute_completion_behavior, $.behavior))),
+        seq(field('name', '-range'), optional(seq(token.immediate('='), field('value', $._command_attribute_range_value)))),
+        seq(field('name', '-count'), token.immediate('='), field('value', alias(token.immediate(/[0-9]+/), $.integer_literal))),
```
Comment (vigoux) [inline grammar.js:476]: Don't we already have `integer_literal` ?

File: grammar.js [hunk 4]
```diff
@@ -395,7 +398,105 @@ module.exports = grammar({
...
+      ),
+
+    _command_attribute_nargs_value: ($) =>
+      choice(
+        alias(token.immediate(/[01]/), $.integer_literal),
+        alias(token.immediate(/[*?+]/), $.pattern_multi),
+      ),
+    _command_attribute_range_value: ($) =>
+      choice(
+        alias(token.immediate(/[0-9]+/), $.integer_literal),
```
Comment (vigoux) [inline grammar.js:467]: `integer_literal` is already present.

**Review by vigoux — CHANGES_REQUESTED** [review]
I am sorry but I think that we could benefit more tests especially for syntax highlighting given the sheer size of this PR.

Regarding the `syntax` problem, please add a test and I'll look into it.

File: queries/highlights.scm (hunk 2 above)
Comment (Fymyte) [reply queries/highlights.scm:143]: The only time there is a node inside a behavior is either for `custom` or `customList` which only accepts function name after the comma

File: grammar.js (hunk 3 above)
Comment (Fymyte) [reply grammar.js:476]: The integer_literal node currently accepts hexadecimal values as well. I'm not sure if it should be inside the integer_literal node that changes should be made or here, but I cannot use it as is here

File: grammar.js (hunk 4 above)
Comment (Fymyte) [reply grammar.js:467]: same

**Comment by Fymyte** [comment]
this
image
seems fine it is highlighted as a function and captured as a function, even though I don't understand why the capture groups appears twice with `:TSHighlightCaptureUnderCursor`
but it fails the test for tree-sitter
image

Here
image
`syntax` is captured as `keyword` twice, but not `constant`, thus highlighting `syntax` as a `keyword` and not `constant`

Test fails in the same way

**Comment by vigoux** [comment]
From what I remember tree-sitter highlighting (the default one) goes the other way that we are do in neovim, that is: the earliest match is the one with the highest precedence.

This may explain the problem here.

**Comment by vigoux** [comment]
After even more digging into the issue, it seems that tree-sitter highlight tests are a bit borked (as I anticipated). I'll try to work on a fix upstream and report status here.

For now I think we can mark these tests as todos.

**Review by vigoux — APPROVED** [review]
Awesome, let's merge that and move on to something else.

**Pull request merged**